x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Categorize.java

```diff
@@ -0,0 +1,159 @@
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+import org.elasticsearch.xpack.ml.aggs.categorization.CategorizationBytesRefHash;
+import org.elasticsearch.xpack.ml.aggs.categorization.CategorizationPartOfSpeechDictionary;
+import org.elasticsearch.xpack.ml.aggs.categorization.TokenListCategorizer;
+import org.elasticsearch.xpack.ml.job.categorization.CategorizationAnalyzer;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.DEFAULT;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isString;
+
+/**
+ * Categorizes text messages.
+ *
+ * This implementation is incomplete and comes with the following caveats:
+ * - it only works correctly on a single node.
+ * - when running on multiple nodes, category IDs of the different nodes are
+ *   aggregated, even though the same ID can correspond to a totally different
+ *   category
+ * - the output consists of category IDs, which should be replaced by category
+ *   regexes or keys
+ *
+ * TODO(jan, nik): fix this
+ */
+public class Categorize extends GroupingFunction implements Validatable {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(
+        Expression.class,
+        "Categorize",
+        Categorize::new
+    );
+
+    private final Expression field;
+
+    @FunctionInfo(returnType = { "integer" }, description = "Categorizes text messages")
+    public Categorize(
+        Source source,
+        @Param(name = "field", type = { "text", "keyword" }, description = "Expression to categorize") Expression field
+    ) {
+        super(source, List.of(field));
+        this.field = field;
+    }
+
+    private Categorize(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(field);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public boolean foldable() {
+        return field.foldable();
+    }
+
+    @Evaluator
+    static int process(
+        BytesRef v,
+        @Fixed(includeInToString = false, build = true) CategorizationAnalyzer analyzer,
+        @Fixed(includeInToString = false, build = true) TokenListCategorizer.CloseableTokenListCategorizer categorizer
+    ) {
+        String s = v.utf8ToString();
+        try (TokenStream ts = analyzer.tokenStream("text", s)) {
+            return categorizer.computeCategory(ts, s.length(), 1).getId();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
We'd talked about allows IOException from a lot of things in ESQL - we do read from files and such. But we never go around to it and I'm not sure we'd do it for `process` anyway. And this one isn't real *anyway* - at least, sort of not. It's throw if the token stream fails to do token stream things, but that's not real IO. Bleh. This is totally fine.